File: src/main/java/frc/team4373/robot/Robot.java
package frc.team4373.robot;

import edu.wpi.first.wpilibj.TimedRobot;
import edu.wpi.first.wpilibj.controller.PIDController;
import edu.wpi.first.wpilibj.controller.ProfiledPIDController;
import edu.wpi.first.wpilibj.geometry.*;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj.trajectory.*;
import edu.wpi.first.wpilibj2.command.CommandScheduler;
import edu.wpi.first.wpilibj2.command.SwerveControllerCommand;
import frc.team4373.robot.commands.camera.CameraCommand;
import frc.team4373.robot.commands.climber.ClimberCommand;
import frc.team4373.robot.commands.drivetrain.DrivetrainCommand;
import frc.team4373.robot.commands.intake.IntakeCommand;
import frc.team4373.robot.commands.shooter.ShooterCommand;
import frc.team4373.robot.input.OI;
import frc.team4373.robot.subsystems.*;

import java.util.List;

public class Robot extends TimedRobot {
    @Override
    public void robotInit() {
        Camera.getInstance().setDefaultCommand(new CameraCommand());
        Climber.getInstance().setDefaultCommand(new ClimberCommand());
        Intake.getInstance().setDefaultCommand(new IntakeCommand());
        Shooter.getInstance().setDefaultCommand(new ShooterCommand());
        Drivetrain.getInstance().setDefaultCommand(new DrivetrainCommand());
        // WheelSpinner.getInstance().setDefaultCommand(new WheelSpinnerCommand());
    }

    @Override
    public void autonomousInit() {
        TrajectoryConfig config = new TrajectoryConfig(RobotMap.MAX_SPEED_METERS_PER_SECOND,
                RobotMap.MAX_ACCELERATION_METERS_PER_SECOND_SQUARED)
                .setKinematics(Drivetrain.getInstance().kinematics);

        Trajectory exampleTrajectory = TrajectoryGenerator.generateTrajectory(
                new Pose2d(0, 0, new Rotation2d()),
                List.of(
                        new Translation2d(1, 1),
                        new Translation2d(2, -1)
                ),
                new Pose2d(3, 0, new Rotation2d()),
                config
        );

        ProfiledPIDController thetaController = new ProfiledPIDController(1, 0, 0,
                RobotMap.THETA_CONTROLLER_CONSTRAINTS);
        thetaController.enableContinuousInput(-Math.PI, Math.PI);

        SwerveControllerCommand swerveControllerCommand = new SwerveControllerCommand(
                exampleTrajectory,
                () -> Drivetrain.getInstance().getPose(),
                Drivetrain.getInstance().kinematics,
                new PIDController(1, 0, 0),
                new PIDController(1, 0, 0), thetaController,
                Drivetrain.getInstance()::setModuleStates,
                Drivetrain.getInstance()
        );

        swerveControllerCommand.schedule();
    }

    @Override
    public void robotPeriodic() {
        CommandScheduler.getInstance().run();

        SmartDashboard.putNumber("shooter/encoder_vel", Shooter.getInstance().getVelocity());
        SmartDashboard.putNumber("shooter/motor1_output",
                Shooter.getInstance().getMotor1PercentOutput());
        SmartDashboard.putNumber("shooter/motor2_output",
                Shooter.getInstance().getMotor2PercentOutput());

        SmartDashboard.putNumber("climber/lift_out",
                Climber.getInstance().getLiftPercentOutput());
        SmartDashboard.putNumber("climber/right_winch_out",
                Climber.getInstance().getRightWinchPercentOutput());
        SmartDashboard.putNumber("climber/left_winch_out",
                Climber.getInstance().getLeftWinchPercentOutput());
        SmartDashboard.putBoolean("climber/bottom_limit_switch",
                Climber.getInstance().getBottomLimitSwitch());
        SmartDashboard.putBoolean("climber/top_limit_switch",
                Climber.getInstance().getTopLimitSwitch());

        SmartDashboard.putNumber("intake/ground_intake_out",
                Intake.getInstance().getGroundMotorPercentOutput());
        SmartDashboard.putNumber("intake/transfer_intake_out",
                Intake.getInstance().getUptakeMotorPercentOutput());
        SmartDashboard.putBoolean("intake/balls_retained",
                Intake.getInstance().getBallsAreRetained());

        SmartDashboard.putNumber("joysticks/throttle_value",
                OI.getInstance().getDriveJoystick().rooGetThrottle());
        SmartDashboard.putNumber("joysticks/input_x",
                OI.getInstance().getDriveJoystick().rooGetX());
        SmartDashboard.putNumber("joysticks/input_y",
                OI.getInstance().getDriveJoystick().rooGetY());
        SmartDashboard.putNumber("joysticks/input_twist",
                OI.getInstance().getDriveJoystick().rooGetTwist());
    }
}

File: src/main/kotlin/frc/team4373/robot/Robot.kt
package frc.team4373.robot

import edu.wpi.first.wpilibj.TimedRobot
import edu.wpi.first.wpilibj.command.Command
import edu.wpi.first.wpilibj.command.Scheduler
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard
import frc.team4373.robot.commands.auton.RamThenShootAuton
import frc.team4373.robot.commands.util.ClearSubsystemCommand
import frc.team4373.robot.input.OI
import frc.team4373.robot.subsystems.Climber
import frc.team4373.robot.subsystems.Drivetrain
import frc.team4373.robot.subsystems.Intake
import frc.team4373.robot.subsystems.Shooter
import frc.team4373.swerve.SwerveDrivetrain

/**
 * The VM is configured to automatically run this class, and to call the
 * functions corresponding to each mode, as described in the TimedRobot
 * documentation. If you change the name of this class or the package after
 * creating this project, you must also update the build.gradle file in the
 * project.
 */
class Robot: TimedRobot() {
    var autonCommand: Command? = null

    /**
     * This function is run when the robot is first started up and should be
     * used for any initialization code.
     *
     *
     * All SmartDashboard fields should be initially added here.
     */
    override fun robotInit() {
        Climber.getInstance()
        Intake.getInstance()
        Shooter.getInstance()
        // WheelSpinner.getInstance();
        Drivetrain.getInstance()
    }

    /**
     * This function is called every robot packet, no matter the mode. Use
     * this for items like diagnostics that you want run during disabled,
     * autonomous, teleoperated, and test.
     *
     *
     * This runs after the mode-specific periodic functions but before
     * LiveWindow and SmartDashboard integrated updating.
     */
    override fun robotPeriodic() {
        SmartDashboard.putNumber("shooter/encoder_vel", Shooter.velocity)
        SmartDashboard.putNumber(
            "shooter/motor1_output",
            Shooter.motor1PercentOutput
        )
        SmartDashboard.putNumber(
            "shooter/motor2_output",
            Shooter.motor2PercentOutput
        )
        SmartDashboard.putNumber(
            "climber/lift_out",
            Climber.getInstance().liftPercentOutput
        )
        SmartDashboard.putNumber(
            "climber/right_winch_out",
            Climber.getInstance().rightWinchPercentOutput
        )
        SmartDashboard.putNumber(
            "climber/left_winch_out",
            Climber.getInstance().leftWinchPercentOutput
        )
        SmartDashboard.putBoolean(
            "climber/bottom_limit_switch",
            Climber.getBottomLimitSwitch()
        )
        SmartDashboard.putBoolean(
            "climber/top_limit_switch",
            Climber.getTopLimitSwitch()
        )
        SmartDashboard.putNumber(
            "intake/ground_intake_out",
            Intake.getInstance().groundMotorPercentOutput
        )
        SmartDashboard.putNumber(
            "intake/transfer_intake_out",
            Intake.getInstance().uptakeMotorPercentOutput
        )
        SmartDashboard.putBoolean(
            "intake/balls_retained",
            Intake.getInstance().ballsAreRetained
        )
        SmartDashboard.putNumber(
            "joysticks/throttle_value",
            OI.getInstance().driveJoystick.rooGetThrottle()
        )
        SmartDashboard.putNumber(
            "joysticks/input_x",
            OI.getInstance().driveJoystick.rooGetX()
        )
        SmartDashboard.putNumber(
            "joysticks/input_y",
            OI.getInstance().driveJoystick.rooGetY()
        )
        SmartDashboard.putNumber(
            "joysticks/input_twist",
            OI.getInstance().driveJoystick.rooGetTwist()
        )


        // SmartDashboard.putString("spinner/color",
        //         WheelSpinner.getInstance().getColor().toString());
        // SmartDashboard.putNumber("spinner/spinner_out",
        //         WheelSpinner.getInstance().getWheelSpinnerPercentOutput());
    }

    /**
     * This function is called once when Sandstorm mode starts.
     */
    override fun autonomousInit() {
        Drivetrain.getInstance().brakeMode = SwerveDrivetrain.BrakeMode.NONE
        autonCommand = RamThenShootAuton()
        (autonCommand as Command).start()
    }

    /**
     * This function is called once when teleoperated mode starts.
     */
    override fun teleopInit() {
        Drivetrain.getInstance().brakeMode = SwerveDrivetrain.BrakeMode.IMPLODE
        // Ensure subsystems are in a known, safe state
        // new ResetSpinnerStateCommand().start();
        ClearSubsystemCommand(Intake.getInstance()).start()
    }

    /**
     * This function is called periodically during Sandstorm mode.
     */
    override fun autonomousPeriodic() {
        Scheduler.getInstance().run()
    }

    /**
     * This function is called periodically during operator control.
     */
    override fun teleopPeriodic() {
        Scheduler.getInstance().run()
    }

    /**
     * This function is called periodically during test mode.
     */
    override fun testPeriodic() {}

    /**
     * This function runs periodically in disabled mode.
     * It is used to verify that the selected auton configuration is legal.
     * <br></br>
     * **UNDER NO CIRCUMSTANCES SHOULD SUBSYSTEMS BE ACCESSED OR ENGAGED IN THIS METHOD.**
     */
    override fun disabledPeriodic() {}
}